sql, row: support INSERT & UPDATE of implicit regional by row partitioning under READ COMMITTED isolation

mw5h opened the pull request:
Under non-Serializable isolations, we need to lock the keys for unique
indexes with implicit partitioning in all partitions to ensure a racing
write doesn't insert a conflicting row into a different partition. Since
there is no support for predicate locks at the KV layer, we fake it by
writing tombstone values to the partitions not being inserted into. This
causes a bit of a write explosion, but it will keep conflicting rows
from being written.

In this patch, we create the infrastructure for writing these tombstones
by modifying the RowHelper to generate a list of keys to lock.

For unique indexes with a single ENUM implicit partition column (e.g.
regional by row), write a tombstone to each partition to ensure
uniqueness when isolation level is not serializable.

This patch also reverts previous work on the insert fast path to write
tombstones because those tombstones are now being written by the row
inserter.

mw5h committed:
sql: support INSERT w/o arbiters under non-serializable isolations

For unique indexes with a single ENUM implicit partition column (e.g.
regional by row), write a tombstone to each partition to ensure
uniqueness when isolation level is not serializable. These tombstones
get write intents, which we use as row locks because the the KV layer
does not currently support locking non-existant rows.

This patch also reverts previous work on the insert fast path to write
tombstones because those tombstones are now being written by the row
inserter.

Part of: #110873
Release note (sql change): REGIONAL BY ROW and PARTITION BY ALL  tables
can now be inserted into under non-serializable isolation levels so long
as there is no ON CONFLICT clause on the statement.

mw5h committed:
sql: support UPDATE on RBR tables under non-serializable isolations

For unique indexes with a single ENUM implicit partition column (e.g.
regional by row), write a tombstone to each partition to ensure
uniqueness when the isolation is not serializable. These tombstones get
write intents, which we are using as row locks because the KV layer does
not currently have the ability to lock non-existant keys.

Fixes: #110873
Release note (sql change): REGIONAL BY ROW and PARTITION ALL BY tables
can now be updated under non-SERIALIZABLE isolation levels.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129837)
<!-- Reviewable:end -->

mw5h commented on pkg/sql/opt_catalog.go:
```diff
@@ -975,13 +975,18 @@ func newOptTable(
+				// of ENUM type (e.g. an implicit RBR column), then we can ensure uniqueness under non-Serializable
+				// isolation levels by writing tombstones
+				canUseTombstones := idx.ImplicitPartitioningColumnCount() == 1 &&
+					catalog.FindColumnByID(desc, idx.GetKeyColumnID(0)).GetType().Family() == types.EnumFamily
```
I'm uncomfortable assuming the RBR column is '0'. Is there a better way to figure this out? At the very least it feels like I should be using a named constant here.

mgartner reviewed (commented):
Great start!

I think we'll want a handful of tests for this, unless we're positive they are already covered (but I don't think we currently run all logic tests under read-committed by default). Tests with unique partial indexes and unique hash-sharded indexes will be particularly interesting.

Also, I found this a bit difficult to follow with the current split of commits. Specifically, introducing functions in one commit that are untested and not used until a follow-up commit makes it harder to understand its purpose and intended behavior.

> Reviewed 4 of 6 files at r1, 18 of 18 files at r3, 9 of 18 files at r4, 30 of 30 files at r6, 6 of 6 files at r7, 18 of 18 files at r8, 10 of 10 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-:-O5yFUy72mUMZ5gydQzn:b-c63bi7)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mw5h, @rytaft, and @stevendanna)

___
*[`pkg/sql/opt_catalog.go` line 980 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yCk7q0TBYMTZKgq6o:-O5yCk7q0TBYMTZKgq6p:bilphl4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a2bde396bcc76cf4a1737f38c05c0d3b5660d7/pkg/sql/opt_catalog.go#L980)):*
> ```Go
> 				// Add unique constraints for implicitly partitioned unique indexes. If there is a single implicit column
> 				// of ENUM type (e.g. an implicit RBR column), then we can ensure uniqueness under non-Serializable
> 				// isolation levels by writing tombstones
> ```

nit: add period to end of sentence
___
*[`pkg/sql/opt_catalog.go` line 981 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yCotK9jUDd2OB2EBv:-O5yCotK9jUDd2OB2EBw:b-6gop86) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a2bde396bcc76cf4a1737f38c05c0d3b5660d7/pkg/sql/opt_catalog.go#L981)):*
> ```Go
> 				// of ENUM type (e.g. an implicit RBR column), then we can ensure uniqueness under non-Serializable
> 				// isolation levels by writing tombstones
> 				canUseTombstones := idx.ImplicitPartitioningColumnCount() == 1 &&
> ```

ImplicitPartitioningColumnCount() will return 2 for a RBR, hash-sharded table. In that case we still want to write a tombstone, I believe. Can we relax this `== 1` to `> 0`?
___
*[`pkg/sql/row/helper.go` line 190 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5y7uaO0jxYpHXmwV6Y:-O5y7uaO0jxYpHXmwV6Z:bdurvoj) ([raw file](https://github.com/cockroachdb/cockroach/blob/db468deb4fdf154d483576d707712ce2cf97cf0d/pkg/sql/row/helper.go#L190)):*
> ```Go
> }
> 
> func (rh *RowHelper) encodeTombstonesForIndex(
> ```

nit: a quick explanation of this function, and/or some comments sprinkled below might be helpful

Also, this function is unused in this commit. That's not necessarily a problem, but it might be a hint that commits should be combined or tests should be added to this commit.
___
*[`pkg/sql/row/helper.go` line 95 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yE99D9wCWttBbMEUp:-O5yE99E9hgdWEXYzHMf:b-ho3en5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cf7958da81d3815c745a843b5b4c2e832b89c55/pkg/sql/row/helper.go#L95)):*
> ```Go
> 
> 	// Used to build tombstones for non-Serializable uniqueness checks
> 	tombstoneCache map[catalog.Index]*tombstoneCacheEntry
> ```

nit: I think this is what we'd call a "scratch", not a "cache", IIUC. The stored keys aren't reused, only the underlying memory is to reduce allocations, correct? Erh... I guess the enum values are cached, so maybe that's not quite right. It might be better to split the cached enums from the scratch. Re: my other comment about returning the tombstones from `encodeIndexes` - if you do that then we might not need the `tombstones [][]byte` at all, right?
___
*[`pkg/sql/row/helper.go` line 216 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yETAc38lxvFYM5PZL:-O5yETAd5MXFNE8i7L9h:bfauumk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cf7958da81d3815c745a843b5b4c2e832b89c55/pkg/sql/row/helper.go#L216)):*
> ```Go
> 	cacheEntry, ok := rh.tombstoneCache[index]
> 	if !ok {
> 		implicitKeys := tree.MakeAllDEnumsInType(colVal.ResolvedType())
> ```

One alternative is to have the optimizer project a column that is an array of tombstone ENUMS for each tombstone index. I'm not sure that necessary makes this simpler, though. I suppose it could perform worse for large inserts because we'll be generating a lot of duplicate arrays of enums.
___
*[`pkg/sql/row/inserter.go` line 125 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yAdBMDtlMekt_J7Rh:-O5yAdBMDtlMekt_J7Ri:bx2nyun) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a2bde396bcc76cf4a1737f38c05c0d3b5660d7/pkg/sql/row/inserter.go#L125)):*
> ```Go
> func (ri *Inserter) writeTombstones(
> 	ctx context.Context, index catalog.Index, b Putter, values []tree.Datum, traceKV bool,
> ) error {
> ```

nit: might be good to add an assertion here verifying the index is unique and/or not an inverted index
___
*[`pkg/sql/row/inserter.go` line 171 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yB__t0Mu2Sn55buqz:-O5yA-bu35S0ALvhMrNQ:bpf22eh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a2bde396bcc76cf4a1737f38c05c0d3b5660d7/pkg/sql/row/inserter.go#L171)):*
> ```Go
> 	// We don't want to insert empty k/v's like this, so we
> 	// set includeEmpty to false.
> 	primaryIndexKey, secondaryIndexEntries, err := ri.Helper.encodeIndexes(
> ```

nit: Did you consider having `encodeIndexes` include the tombstones to write? That would simplify things below.
___
*[`pkg/sql/row/inserter.go` line 213 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yAINM0Iuc284YfsZK:-O5yAINM0Iuc284YfsZL:b-ht2dtx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a2bde396bcc76cf4a1737f38c05c0d3b5660d7/pkg/sql/row/inserter.go#L213)):*
> ```Go
> 
> 			if !wrote {
> 				continue
> ```

nit: I think this would be easier to grok:

```go
if len(entries) > ) {
    if err := ri.writeTombstones(...) err != nil {
        return err
    }
}
```
___
*[`pkg/sql/row/inserter.go` line 215 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yAmsw87-eFXX3HvMo:-O5yAmsw87-eFXX3HvMp:bq3gd2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a2bde396bcc76cf4a1737f38c05c0d3b5660d7/pkg/sql/row/inserter.go#L215)):*
> ```Go
> 				continue
> 			}
> 			if err := ri.writeTombstones(ctx, index, b, values, traceKV); err != nil {
> ```

nit: We should mention here how this works with unique partial indexes, since it's not obviously respecting `pm.IgnoreForPut`, e.g., "if a row does not satisfy the partial index predicate, then it will have empty `entries` and therefore no tombstone will be written."


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
This is looking great! Agree that you need more tests. I mostly left a bunch of nits.

Another nit is that the PR title should be prefixed with the affected package(s) similar to commit messages

> Reviewed 4 of 6 files at r1, 18 of 18 files at r3, 18 of 18 files at r4, 9 of 10 files at r5, 8 of 30 files at r6, 4 of 6 files at r7, 10 of 18 files at r8, 10 of 10 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-:-O5yVaVq3ZNQsynajxXF:b-fsdolt)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mw5h and @stevendanna)

___
*[`-- commits` line 30 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5xQAFqDQy5Gz3zyzu7):*
nit: What do you mean by "implicit regional by row" tables? Maybe you mean "REGIONAL BY ROW and PARTITION ALL BY" tables?
___
*[`-- commits` line 42 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5xdbsv5K6jsWxBofhI):*
ditto on "implicit regional by row"
___
*[`-- commits` line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yUGtj6oABUajIMWep):*
I think you've pulled an unrelated commit into your PR
___
*[`pkg/sql/opt_catalog.go` line 982 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/129837#gh-1735350855:-O5xQjip4PiZTnBdZCrU:bcsula1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac374753499cd61fed7d325da9e5b78ba8140b7b/pkg/sql/opt_catalog.go#L982)):*
<details><summary><i>Previously, mw5h (Matt White) wrote…</i></summary><blockquote>

I'm uncomfortable assuming the RBR column is '0'. Is there a better way to figure this out? At the very least it feels like I should be using a named constant here.
</blockquote></details>

By design, the partitioning columns have to be first in the index. So I think what you're doing is fine. I don't think we want a named constant, since there could be multiple partition columns. In general we prefer the optimizer to work with primitives and not at the level of RBR anyway (although there are some places where RBR has already leaked in, unfortunately).
___
*[`pkg/sql/opt_catalog.go` line 980 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5xZQRJEhwgCC9Udfad:-O5xZQRJEhwgCC9Udfae:b-7tj4z5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a2bde396bcc76cf4a1737f38c05c0d3b5660d7/pkg/sql/opt_catalog.go#L980)):*
> ```Go
> 				// Add unique constraints for implicitly partitioned unique indexes. If there is a single implicit column
> 				// of ENUM type (e.g. an implicit RBR column), then we can ensure uniqueness under non-Serializable
> 				// isolation levels by writing tombstones
> ```

nit: comments should be 80 characters wide an end with a period.
___
*[`pkg/sql/opt/cat/table.go` line 365 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5x_1ou7WDxwoH4RAZo:-O5x_1ou7WDxwoH4RAZp:b-dqb4l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a2bde396bcc76cf4a1737f38c05c0d3b5660d7/pkg/sql/opt/cat/table.go#L365)):*
> ```Go
> 	WithoutIndex() bool
> 
> 	CanUseTombstones() bool
> ```

nit: add a comment
___
*[`pkg/sql/opt/exec/explain/emit.go` line 40 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5xbzMnCuQAWU7kctpW:-O5xbzMnCuQAWU7kctpX:b-dqb4l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a2bde396bcc76cf4a1737f38c05c0d3b5660d7/pkg/sql/opt/exec/explain/emit.go#L40)):*
> ```Go
> }
> 
> func joinNames(table cat.Table, ords cat.IndexOrdinals, sep string) (string, bool) {
> ```

nit: add a comment
___
*[`pkg/sql/row/helper.go` line 69 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5YGSp07w24SGIXrtfM:-O5YGSp07w24SGIXrtfN:b-ti0323) ([raw file](https://github.com/cockroachdb/cockroach/blob/db468deb4fdf154d483576d707712ce2cf97cf0d/pkg/sql/row/helper.go#L69)):*
> ```Go
> )
> 
> type tombstoneCacheEntry struct {
> ```

nit: add a comment to explain what this is
___
*[`pkg/sql/row/helper.go` line 81 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5YGNAs1XXkgItLb7QO:-O5YGNAt1vmy430xaWic:b-dqb4l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/db468deb4fdf154d483576d707712ce2cf97cf0d/pkg/sql/row/helper.go#L81)):*
> ```Go
> 	// Secondary indexes.
> 	Indexes                    []catalog.Index
> 	UniqueWithTombstoneIndexes intsets.Fast
> ```

nit: add a comment
___
*[`pkg/sql/row/helper.go` line 200 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5YH4g6AC0oqJpz52Oi:-O5YH4g6AC0oqJpz52Oj:bhciark) ([raw file](https://github.com/cockroachdb/cockroach/blob/db468deb4fdf154d483576d707712ce2cf97cf0d/pkg/sql/row/helper.go#L200)):*
> ```Go
> 	}
> 
> 	colIdx, ok := colIDtoRowIndex.Get(index.GetKeyColumnID((0)))
> ```

nit: explain what 0 means here (i.e. this is the index partition column)

Also, it's a bit odd that something called `colIdtoRowIndex` returns something called `colIdx`. Is this correct?
___
*[`pkg/sql/row/helper.go` line 204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5xNXl0597gEPi_QnB-:-O5xNXl0597gEPi_QnB0:bsa23y) ([raw file](https://github.com/cockroachdb/cockroach/blob/db468deb4fdf154d483576d707712ce2cf97cf0d/pkg/sql/row/helper.go#L204)):*
> ```Go
> 		return nil, nil
> 	}
> 	colVal := values[colIdx]
> ```

should we assert that colVal has type `*tree.DEnum`? (i.e. return `errors.AssertionFailedf(...)` if not)
___
*[`pkg/sql/rowenc/index_encoding.go` line 1289 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5xPk-W-_-_D16Iq97Q:-O5xPk-X06GruXGMp4ef:bx2a4a2) ([raw file](https://github.com/cockroachdb/cockroach/blob/db468deb4fdf154d483576d707712ce2cf97cf0d/pkg/sql/rowenc/index_encoding.go#L1289)):*
> ```Go
> }
> 
> func EncodeSecondaryIndexKey(
> ```

nit: add a comment for this function
___
*[`pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_read_committed` line 120 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yVH3n5NHN207MwsrR:-O5yVH3n5NHN207MwsrS:b-ylzudg) ([raw file](https://github.com/cockroachdb/cockroach/blob/945c725f1b6f2c92cb4896622828071a2e55ecf4/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_read_committed#L120)):*
> ```
> ON CONFLICT (mascot) DO NOTHING
> 
> # Temporary until ON CONFLICT works
> ```

nit: prefix this with `TODO(mw5h): `


<!-- Sent from Reviewable.io -->

mw5h reviewed (commented):
pkg/ccl/logictestccl/testdata/logic_test/regional_by_row_read_committed provides some tests, but I can probably expand upon them.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-:-O5yzKNt2LoaDvvIiYUM:b-dlf7el)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rytaft, and @stevendanna)

___
*[`-- commits` line 30 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5xQAFqDQy5Gz3zyzu7):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: What do you mean by "implicit regional by row" tables? Maybe you mean "REGIONAL BY ROW and PARTITION ALL BY" tables?
</blockquote></details>

Yes, that's it. I meant implicitly partitioned tables with an enumerated partition key.
___
*[`pkg/sql/opt_catalog.go` line 981 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yCotK9jUDd2OB2EBv:-O5z-VG85_LowvDkJyHM:b-nkpola) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a2bde396bcc76cf4a1737f38c05c0d3b5660d7/pkg/sql/opt_catalog.go#L981)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

ImplicitPartitioningColumnCount() will return 2 for a RBR, hash-sharded table. In that case we still want to write a tombstone, I believe. Can we relax this `== 1` to `> 0`?
</blockquote></details>

I'm not sure that's safe if there are multiple implicit partition columns. If this check is overly permissive, the result will be dropped uniqueness constraints, so I want to be careful. Assuming that the shard column value is the same for every partition, I think it would be safe to allow this value to be '2' if the index is sharded.
___
*[`pkg/sql/row/helper.go` line 69 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5YGSp07w24SGIXrtfM:-O5z0Mnb7ZstV_DBWH9x:b-ahfhlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/db468deb4fdf154d483576d707712ce2cf97cf0d/pkg/sql/row/helper.go#L69)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: add a comment to explain what this is
</blockquote></details>

Heh, can you tell that we tended to discourage comments in the Clustrix code base?
___
*[`pkg/sql/row/helper.go` line 204 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5xNXl0597gEPi_QnB-:-O5z0oOK0_4dNBMBFe9k:b-xxyxcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/db468deb4fdf154d483576d707712ce2cf97cf0d/pkg/sql/row/helper.go#L204)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should we assert that colVal has type `*tree.DEnum`? (i.e. return `errors.AssertionFailedf(...)` if not)
</blockquote></details>

I love assertions.
___
*[`pkg/sql/row/helper.go` line 95 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yE99D9wCWttBbMEUp:-O5z0RmLCjjXgsmhZ3uj:buj8dwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cf7958da81d3815c745a843b5b4c2e832b89c55/pkg/sql/row/helper.go#L95)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: I think this is what we'd call a "scratch", not a "cache", IIUC. The stored keys aren't reused, only the underlying memory is to reduce allocations, correct? Erh... I guess the enum values are cached, so maybe that's not quite right. It might be better to split the cached enums from the scratch. Re: my other comment about returning the tombstones from `encodeIndexes` - if you do that then we might not need the `tombstones [][]byte` at all, right?
</blockquote></details>

This started life as just a cache of the enumerated values and then got expanded, but yeah, I can take a look at that.
___
*[`pkg/sql/row/helper.go` line 216 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yETAc38lxvFYM5PZL:-O5z1H-pB_5N2UpPVIdr:b-p35qtp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3cf7958da81d3815c745a843b5b4c2e832b89c55/pkg/sql/row/helper.go#L216)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

One alternative is to have the optimizer project a column that is an array of tombstone ENUMS for each tombstone index. I'm not sure that necessary makes this simpler, though. I suppose it could perform worse for large inserts because we'll be generating a lot of duplicate arrays of enums.
</blockquote></details>

I don't know that it would be any worse and might make EXPLAIN output better.
___
*[`pkg/sql/row/inserter.go` line 171 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yB__t0Mu2Sn55buqz:-O5z1Q_0AQNItGdGvCKF:b-tzzzfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a2bde396bcc76cf4a1737f38c05c0d3b5660d7/pkg/sql/row/inserter.go#L171)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Did you consider having `encodeIndexes` include the tombstones to write? That would simplify things below.
</blockquote></details>

An early version of this patch did exactly that, but then I got to UPDATE and needed to conditionally write tombstones only when an index was changed.
___
*[`pkg/sql/row/inserter.go` line 213 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yAINM0Iuc284YfsZK:-O5z1uUB0-UbI9lgz1J9:btd51a6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a2bde396bcc76cf4a1737f38c05c0d3b5660d7/pkg/sql/row/inserter.go#L213)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: I think this would be easier to grok:

```go
if len(entries) > ) {
    if err := ri.writeTombstones(...) err != nil {
        return err
    }
}
```
</blockquote></details>

I agree.
___
*[`pkg/sql/row/inserter.go` line 215 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yAmsw87-eFXX3HvMo:-O5z25cQ6ePS8aK3c5EU:bebrt1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a2bde396bcc76cf4a1737f38c05c0d3b5660d7/pkg/sql/row/inserter.go#L215)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: We should mention here how this works with unique partial indexes, since it's not obviously respecting `pm.IgnoreForPut`, e.g., "if a row does not satisfy the partial index predicate, then it will have empty `entries` and therefore no tombstone will be written."
</blockquote></details>

Will do


<!-- Sent from Reviewable.io -->

mw5h reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129837)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rytaft, and @stevendanna)

___
*[`-- commits` line 42 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5xdbsv5K6jsWxBofhI):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

ditto on "implicit regional by row"
</blockquote></details>

Done.
___
*[`-- commits` line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5yUGtj6oABUajIMWep):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think you've pulled an unrelated commit into your PR
</blockquote></details>

Done.
___
*[`pkg/sql/row/helper.go` line 200 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O5YH4g6AC0oqJpz52Oi:-O5zRXCD5oQ523nAVPuU:b-yz2jz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/db468deb4fdf154d483576d707712ce2cf97cf0d/pkg/sql/row/helper.go#L200)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: explain what 0 means here (i.e. this is the index partition column)

Also, it's a bit odd that something called `colIdtoRowIndex` returns something called `colIdx`. Is this correct?
</blockquote></details>

That was lifted from encodeSecondaryIndex(), but I'll change it.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Nice! Expanding the tests would be good. I think we're also missing coverage of hash sharded RBR tables and sharded/regular PARTITION ALL BY tables.

> Reviewed 10 of 12 files at r11, 10 of 18 files at r12, 10 of 10 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-:-O6hBlZ7BwqWIUIYq8Fd:btqlnsd)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @mw5h, and @stevendanna)

___
*[`pkg/sql/row/helper.go` line 70 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O6hACpP2PPgwm3XAL5D:-O6hACpQ1HJjXxLzO-tO:b-7kjrl7) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec0d28872cc2076b4b9cbe8ffe61f03b00c6dad/pkg/sql/row/helper.go#L70)):*
> ```Go
> )
> 
> // Per-index data for writing tombstones to enforce a uniqueness constraint
> ```

nit: end comments with a period, here and below.
___
*[`pkg/sql/row/helper.go` line 184 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O6hAQII57nEdEhoqhLa:-O6hAQII57nEdEhoqhLb:b-5y8e5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec0d28872cc2076b4b9cbe8ffe61f03b00c6dad/pkg/sql/row/helper.go#L184)):*
> ```Go
> 
> // encodePrimaryIndex encodes the primary index key.
> // encodePrimaryIndexKey encodes the primary index key.
> ```

nit: remove old comment above this one.


<!-- Sent from Reviewable.io -->

mw5h commented:
@rytaft @mgartner I added a test for implicit partitioning, but not for sharded indexes. It turns out that those already worked, so I just deleted the logic around them in opt_catalog.go.

rytaft reviewed (commented):
> [@rytaft](https://github.com/rytaft) [@mgartner](https://github.com/mgartner) I added a test for implicit partitioning, but not for sharded indexes. It turns out that those already worked, so I just deleted the logic around them in opt\_catalog.go.

The new tests look good! Are there tests somewhere that show the hash sharded indexes work with RC?

> Reviewed 4 of 6 files at r15, 12 of 21 files at r16, 11 of 11 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-:-O82hqDC1wQqw7QuDxyF:bwl9b2i)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @mw5h, and @stevendanna)

___
*[`pkg/sql/row/inserter.go` line 139 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O82ivfb8Q-PzzM23Od2:-O82ivfb8Q-PzzM23Od3:b-ecj9oo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6500697c1da895374ba670154024915dcd070db9/pkg/sql/row/inserter.go#L139)):*
> ```Go
> 		k := roachpb.Key(keys.MakeFamilyKey(tombstone, 0 /* famID */))
> 		if traceKV {
> 			log.VEventfDepth(ctx, 1, 2, "CPut %s -> nil (tombstone)", k)
> ```

How about adding a kv trace test that shows this output using `SHOW TRACE FOR SESSION`?
___
*[`pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit_read_committed` line 54 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O82gknjEC9kFinoaguz:-O82gknjEC9kFinoagv-:bsx3g3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/6500697c1da895374ba670154024915dcd070db9/pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit_read_committed#L54)):*
> ```
> 
> query T
> SELECT create_statement FROM [SHOW CREATE TABLE t]
> ```

nit: this is a duplicate of the one above
___
*[`pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit_read_committed` line 151 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O82iFnH1GV0JPnwmtfI:-O82iFnI3YYxQ81Ui_tr:b-hwlk57) ([raw file](https://github.com/cockroachdb/cockroach/blob/6500697c1da895374ba670154024915dcd070db9/pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit_read_committed#L151)):*
> ```
> 
> statement error pgcode 23505 pq: duplicate key value violates unique constraint "t_c_key"
> UPDATE t SET c = 4 WHERE pk = 2
> ```

Nice. How about also adding a test showing that if you use multiple partitioning columns or non-enum columns you get an appropriate error?


<!-- Sent from Reviewable.io -->

mw5h committed:
row: refactor some confusing function & variable names

The name colIDtoRowIdx was creating confusion in another patch. 'Idx' is
overloaded in the context of a database, so change that to the more
descriptive 'Position'. We also change 'encodePrimaryIndex()' to
'encodePrimaryIndexKey()' to better describe what we're doing and make
code more obvious.

Part of: #129837
Release note: None

mw5h committed:
row: add RowHelper support for writing tombstones on implicitly partitioned indexes

Under non-Serializable isolations, we need to lock the keys for unique
indexes with implicit partitioning in all partitions to ensure a racing
write doesn't insert a conflicting row into a different partition. Since
there is no support for predicate locks at the KV layer, we fake it by
writing tombstone values to the partitions not being inserted into. This
causes a bit of a write explosion, but it will keep conflicting rows
from being written.

In this patch, we create the infrastructure for writing these tombstones
by modifying the RowHelper to generate a list of keys to lock.

Part of: #110873
Release note: None

mw5h commented:
RFAL

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice work! Just one small nit remaining

(one request for the future with code reviews: in addition to posting "RFAL", please respond "Done." to messages that you've addressed so it's easy to confirm and resolve them.)

> Reviewed 3 of 6 files at r20, 14 of 23 files at r21, 11 of 11 files at r22, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-:-O9BG0ct7iE21JC9ApuK:b-vc7cyq)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @herkolategan, @mgartner, @mw5h, @stevendanna, @vidit-bhat, @wenyihu6, and @xinhaoz)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit_read_committed` line 12 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O9BF3FOERddpe409X8X:-O9BF3FOERddpe409X8Y:b-ye8qx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/71a7f579e5acc98a4f1c52523d24f56dec1eb208/pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit_read_committed#L12)):*
> ```
> SET SESSION CHARACTERISTICS AS TRANSACTION ISOLATION LEVEL READ COMMITTED
> 
> statement ok
> ```

nit: add a comment explaining what these two tests are showing (multiple partitioning columns and non-enum columns). And/or consider improving the error message to mention the specific reason these checks are unsupported.


<!-- Sent from Reviewable.io -->

mw5h reviewed (commented):
_pkg/ccl/logictestccl/testdata/logic_test/hash_sharded_index_read_committed_

It's pretty basic.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-:-O8909uB2tU4l9CUqscK:b-u0lanr)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @herkolategan, @mgartner, @rytaft, @stevendanna, @vidit-bhat, @wenyihu6, and @xinhaoz)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit_read_committed` line 12 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-O9BF3FOERddpe409X8X:-O9BPelw65MZ2Ty_GP9r:b-v2wwnl) ([raw file](https://github.com/cockroachdb/cockroach/blob/71a7f579e5acc98a4f1c52523d24f56dec1eb208/pkg/ccl/logictestccl/testdata/logic_test/partitioning_implicit_read_committed#L12)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: add a comment explaining what these two tests are showing (multiple partitioning columns and non-enum columns). And/or consider improving the error message to mention the specific reason these checks are unsupported.
</blockquote></details>

I added the comments. The place where the error code is generated doesn't know why we can't use tombstones, so there isn't an easy way to make a better error message. That code is changing a bit on my arbiters branch, so maybe I can improve the error message there.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 107 files at r36, 14 of 23 files at r37, 11 of 11 files at r38, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129837#-:-O9BfiqU2G_dUKEYus-P:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @herkolategan, @mgartner, @stevendanna, @vidit-bhat, @wenyihu6, and @xinhaoz)



<!-- Sent from Reviewable.io -->

mw5h commented:
bors r+
